test/zzzz-bulk_rejection.sh

```diff
@@ -0,0 +1,259 @@
+
+# turn on debug output
+cat $f_cm | \
+    sed -e 's,@include configs.d/openshift/system.conf,<system>\
+      log_level debug\
+    </system>,' | oc replace --force -f -
+
+oc set env ds/logging-fluentd DEBUG=true
+
+# save current es settings
+
+es_settings=$( mktemp )
+curl_es $espod /_cluster/settings | jq . > $es_settings
+
+# change bulk queue_size and size to 1 to make it easy to overload
+curl_es $espod /_cluster/settings -XPUT -d '{
+    "transient" : {
+        "threadpool.bulk.queue_size" : 1,
+        "threadpool.bulk.size": 1
+    }
+}' | jq . | artifact_out
+
+# check settings
+curl_es $espod /_cat/thread_pool?v\&h=bc,br,ba,bq,bs | artifact_out
+
+# create a really large bulk index request json file:
+bulktestjson=$( mktemp )
+python -c 'import sys
+for ii in xrange(0,int(sys.argv[1])):
+    print """
+{{"index":{{"_index":"bulkindextest","_type":"bulkindextest"}}}}
+{{"field0":"value value value {0}"}}
+""".format(ii)
+' 100000 > $bulktestjson
+wc $bulktestjson | artifact_out
+ls -al $bulktestjson | artifact_out
+
+# start curl doing many bulk index ops
+bulkdonefile=$( mktemp )
+echo $bulkdonefile | artifact_out
+do_curl_bulk_index() {
+    local parallel_curls=6
+    local ii
+    local bulkpids=""
+    for ii in $( seq 1 $parallel_curls ) ; do
+        while [ ! -s $bulkdonefile -a -n "${bulktestjson:-}" -a -f "${bulktestjson:-}" ] ; do
+            cat $bulktestjson | curl_es_input $espod /_bulk -XPOST --data-binary @- > /dev/null
+        done & bulkpids="$bulkpids $!"
+    done
+    curl_es $espod /_cat/thread_pool?v\&h=bc,br,ba,bq,bs | artifact_out
+    wait $bulkpids
+    curl_es $espod /_cat/thread_pool?v\&h=bc,br,ba,bq,bs | artifact_out
+}
+
+do_curl_bulk_index & curlpid=$!
+# wait for elasticsearch to report bulk index rejections
+os::cmd::try_until_not_text "curl_es $espod /_cat/thread_pool?h=br" "^0\$"
+
+# restart fluentd to make sure the logs are clear
+os::cmd::try_until_text "oc get pods -l component=fluentd" "^logging-fluentd-.* Running "
+fpod=$( get_running_pod fluentd )
+oc delete pod --force $fpod
+os::cmd::try_until_failure "oc describe pod $fpod"
+sleep 1
+os::cmd::try_until_text "oc get pods -l component=fluentd" "^logging-fluentd-.* Running "
+fpod=$( get_running_pod fluentd )
+
+# wait for BulkIndexQueueFull errors in fluentd log
+flog=/var/log/fluentd.log
+os::cmd::try_until_success "sudo grep -q BulkIndexQueueFull /var/log/fluentd.log" $(( 240 * second ))
+
+# write some messages
+uuid_es=$( openssl rand -hex 64 )
+uuid_es_ops=$( openssl rand -hex 64 )
+
+wait_for_fluentd_ready
+count=40
+countops=500
+os::log::info Adding $count project log records and $countops operations log records . . .
+starttime=$( date +%s )
+# not sure why, but it seems the operations messages get sent
+# to es much faster - so add more of them to see if we can get
+# them to be rejected
+for jj in $( seq 1 $count ) ; do
+    add_test_message "$uuid_es-$jj"
+    os::log::debug added es message $uuid_es-$jj
```

Comment:
Do I get it correctly that calling 40x times the `add_test_message` should result into BulkRequestRejection? Is it because we changed bulk thread configuration in ES prior that? In other words is it correct to assume that fluentd will produce more than one bulk request here?

Reply:
> Do I get it correctly that calling 40x times the add_test_message should result into BulkRequestRejection?

The intention is that fluentd gets several bulk index rejection exceptions when trying to send these test messages.  I picked 40 because it seemed like a good number based on repeated trial and error.

>  Is it because we changed bulk thread configuration in ES prior that?

That, and the fact that there are multiple bulk index requests from curl happening at the same time.  That causes fluentd to fail its bulk index requests.

> In other words is it correct to assume that fluentd will produce more than one bulk request here?

There is no way to guarantee that, but in my testing, there were always at least 4-5 bulk index rejections reported by fluentd.

Reply:
>> Is it because we changed bulk thread configuration in ES prior that?

> That, and the fact that there are multiple bulk index requests from curl happening at the same time. That causes fluentd to fail its bulk index requests.

That means the `add_test_message` is async? Or other curl requests interfere with this?

Reply:
> the add_test_message is async?

Not sure what you mean by `async` in this context.  `add_test_message` uses `curl` to hit kibana with an unknown URL which causes it to emit a `404 not found` to its log, which is then picked up by docker, then fluentd, which tries to submit a bulk index request to Elasticsearch with this record and usually many more.  `add_test_message` returns immediately - it does not run as a background job.

This function starts 6 curl processes running in parallel in the background, all hitting Elasticsearch with as many bulk index requests as possible.  https://github.com/openshift/origin-aggregated-logging/pull/933/files#diff-1e94ccb79adaf3b43bc5ad62f61efd0bR96  All of these curl processes cause Elasticsearch to become overloaded and start rejecting bulk index requests.